src/Pangraph/Internal/Graph.hs

```diff
@@ -0,0 +1,206 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Pangraph.Internal.Graph (
+-- Abstract Types
+Pangraph,
+MalformedEdge,
+Edge,
+Vertex,
+Attribute, -- A type alias for (Key, Value)
+Key,
+Value,
+VertexID,
+EdgeID,
+
+-- Constructors
+makePangraph,
+makeEdge,
+makeVertex,
+
+-- Pangraph Getters
+edges,
+vertices,
+mapEdges,
+mapVertices,
+vertexByID,
+edgeByID,
+
+-- -- Operations on Pangraph
+-- addEdge,
+-- updateEdge,
+-- insertVertex,
+-- removeEdge,
+-- removeVertex,
+
+-- Getters on Vertex and Edge
+edgeAttributes,
+vertexAttributes,
+edgeEndpoints,
+edgeID,
+vertexID,
+lookupVertexValues,
+lookupEdgeValues,
+vertexContainsKey,
+edgeContainsKey,
+
+-- Utility Operations
+vertexToAssocList,
+edgeToAssocList
+
+) where
+
+import Data.Map.Strict (Map)
+import Data.List (intercalate)
+import Data.Maybe (catMaybes, mapMaybe, fromMaybe)
+import Data.ByteString.Char8 (pack, unpack)
+import qualified Data.Map.Strict                  as Map
+import qualified Data.ByteString                  as BS
+
+data Pangraph = Pangraph
+  { vertices' :: Map VertexID Vertex
+  , edges' :: Map EdgeID Edge
+  , nextEdge' :: EdgeID
+  } deriving (Eq)
+data Vertex = Vertex
+  { vertexID' :: VertexID
+  , vertexAttributes' :: [Attribute]
+  } deriving (Eq)
+data Edge = Edge
+  { edgeID' :: Maybe EdgeID
+  , edgeAttributes' :: [Attribute]
+  , endpoints' :: (Vertex, Vertex)
+  } deriving (Eq)
+
+type VertexID = BS.ByteString
+type EdgeID = Int
+type Attribute = (Key, Value)
+type Key = BS.ByteString
+type Value = BS.ByteString
+type MalformedEdge = (Edge, (Maybe Vertex, Maybe Vertex))
```

Comment:
Do you really need a separate type synonym for malformed edges? I think you are giving too much attention to malformed edges. It's fine to simply return the `EdgeID` of the first malformed edge, so I suggest you drop this functionality, because it makes the API more complex.